circuit-foundation/src/commonMain/kotlin/com/slack/circuit/foundation/AnsweringNavigator.kt

```diff
@@ -0,0 +1,134 @@
+  fallbackNavigator: Navigator,
+  resultType: KClass<T>,
+  block: suspend CoroutineScope.(result: T) -> Unit,
+): GoToNavigator {
+  val backStack = LocalBackStack.current ?: return fallbackNavigator
+  return rememberAnsweringNavigator(backStack, resultType, block)
+}
+
+/**
+ * A reified version of [rememberAnsweringNavigator]. See documented overloads of this function for
+ * more information.
+ */
+@Composable
+public inline fun <reified T : PopResult> rememberAnsweringNavigator(
+  backStack: BackStack<out BackStack.Record>,
+  noinline block: suspend CoroutineScope.(result: T) -> Unit,
+): GoToNavigator {
+  return rememberAnsweringNavigator(backStack, T::class, block)
+}
+
+/**
+ * Returns a [GoToNavigator] that answers with the given [resultType] using the given [backStack].
+ *
+ * Handling of the result type [T] should be handled in the [block] parameter and is guaranteed to
+ * only be called _at most_ once. It may never be called if the navigated screen never pops with a
+ * result (of equivalent type) back.
+ *
+ * Note that [resultType] is a simple instance check, so subtypes may also be valid answers.
+ *
+ * ## Example
+ *
+ * ```kotlin
+ * val pickPhotoNavigator = rememberAnsweringNavigator(backStack, PickPhotoScreen.Result::class) { result: PickPhotoScreen.Result ->
+ *   // Do something with the result!
+ * }
+ *
+ * return State(...) { event ->
+ *   when (event) {
+ *     is PickPhoto -> pickPhotoNavigator.goTo(PickPhotoScreen)
+ *   }
+ * }
+ *
+ * // In PickPhotoScreen
+ * navigator.pop(PickPhotoScreen.Result(...))
+ * ```
+ */
+@Composable
+public fun <T : PopResult> rememberAnsweringNavigator(
+  backStack: BackStack<out BackStack.Record>,
+  resultType: KClass<T>,
+  block: suspend CoroutineScope.(result: T) -> Unit,
+): GoToNavigator {
+  val currentBackStack by rememberUpdatedState(backStack)
+  val currentResultType by rememberUpdatedState(resultType)
+
+  // Top screen at the start, so we can ensure we only collect the result if
+  // we've returned to this screen
+  val initialRecordKey = rememberSaveable {
+    currentBackStack.topRecord?.key ?: error("Navigator must have a top screen at start.")
+  }
+
+  // Key for the resultKey, so we can track who owns this requested result
+  val key = rememberSaveable { uuid4().toString() }
+
+  // Current top record of the navigator
+  val currentTopRecordKey by remember { derivedStateOf { currentBackStack.topRecord!!.key } }
+
+  // Track whether we've actually gone to the next record yet
+  var launched by rememberSaveable { mutableStateOf(false) }
+
+  // Collect the result if we've launched and now returned to the initial record
+  if (launched && currentTopRecordKey == initialRecordKey) {
+    val record = currentBackStack.topRecord!!
+    LaunchedEffect(key) {
+      val result = record.awaitResult(key) ?: return@LaunchedEffect
+      if (currentResultType.isInstance(result)) {
+        @Suppress("UNCHECKED_CAST") block(result as T)
+      }
```

Comment:
Should `launched` get reset after handling the result?

Reply:
You're right, though this is taking me down an interesting rabbit hole to set up in a test that may have revealed another bug around not clearing record state when popping through one unanswered record to an answered one. Working on it

Reply:
Ok, I've added a pretty good test for this scenario (`mixedAnswers()`) and added this reset after we result a result (or null), but I'm actually unable to create a case where it would fail without it. Left my logging infra in temporarily and pushed. Do you wanna take a look and see if there's anything I'm missing?

Reply:
Can't really break it either, test looks good 👍🏻